notebooks/demo.jl

```diff
@@ -0,0 +1,636 @@
+"""
+
+# ╔═╡ 8f207356-c29a-4762-a7ce-011b8919f667
+autodiff(square, 1.0)
+
+# ╔═╡ 60d880ac-b534-41b6-a868-3aead8cdf827
+md"""
+Default activity for values is `Const`
+"""
+
+# ╔═╡ b8b856f1-9bd6-40fc-a141-280b0f8a34ff
+autodiff(square, Const(1.0))
+
+# ╔═╡ cf180dad-90a5-4380-9b74-7725dcaf0e73
+autodiff(square, Active(1.0))
+
+# ╔═╡ 6bf01d28-9c5e-4220-bc42-bf446ee7d807
+md"""
+Support for mutations
+"""
+
+# ╔═╡ 7dfe8143-bd09-4046-8141-bd51e12a3e5d
+function cube(y, x)
+	y[] = x[]^3
+	return nothing
+end
+
+# ╔═╡ 124e04fe-9393-44de-a556-52fd5d1cf9db
+let
+	x = Ref(4.0)
+	y = Ref(0.0)
+	cube(y, x)
+	y[]
+end
+
+# ╔═╡ bda1fecc-2499-4deb-bc4d-988e29887929
+md"""
+In order to calculate the gradient of `x`, we have to propagate `1.0` into the
+shadow `dy`.
+"""
+
+# ╔═╡ fb55cc7a-dc83-4d42-87d3-c1ab3450ca28
+let
+	x = Ref(4.0)
+	dx = Ref(0.0)
+	
+	y = Ref(0.0)
+	dy = Ref(1.0)
+	
+	autodiff(cube, Duplicated(y, dy), Duplicated(x, dx))
+	y[], dy[], x[], dx[]
+end
+
+# ╔═╡ 59c51921-a7ce-4014-827f-a460ea0e2e7e
+md"""
+If we do not care about the return value of our function and are only interested in calculating the gradient we can use `DuplicatedNoNeed`.
+"""
+
+# ╔═╡ 258c181e-52f3-4a8f-b2ca-6713c7137766
+let
+	x = Ref(4.0)
+	dx = Ref(0.0)
+	
+	y = Ref(0.0)
+	dy = Ref(1.0)
+	
+	autodiff(cube, Enzyme.DuplicatedNoNeed(y, dy), Duplicated(x, dx))
+	y[], dy[], x[], dx[]
+end
+
+# ╔═╡ 39279922-2994-4d2a-83c0-fead3083f247
+# Enzyme.Compiler.enzyme_code_llvm(cube,
+# 	Tuple{Enzyme.DuplicatedNoNeed{Base.RefValue{Float64}}, 
+# 	Duplicated{Base.RefValue{Float64}}})
+
+# ╔═╡ 5e85ad6a-6508-4bf0-b25b-a64ba65556f5
+md"""
+Let's differentiate through some control flow
+"""
+
+# ╔═╡ 6826dc6e-2e9c-4dcc-89c4-cc98ca03cab0
+# Taylor series for `-log(1-x)`
+# eval at -log(1-1/2) = -log(1/2)
+function taylor(f::T, N) where T
+    g = zero(T)
+    for i in 1:N
+        g += f^i / i
+    end
+    return g
+end
+
+# ╔═╡ 58ca6247-082e-4f66-947b-29181c3ebedf
+autodiff(taylor, Active(0.5), Const(10^7))
+
+# ╔═╡ e21a7977-73b2-4c6a-90b7-3c9dbc04e613
+md"""
+A custom matrix multiply
+"""
+
+# ╔═╡ 21858576-c2a0-474f-9aa7-d14ee7f7210c
+function mymul!(R, A, B)
+    @assert axes(A,2) == axes(B,1)
+    @inbounds @simd for i in eachindex(R)
+        R[i] = 0
+    end
+    @inbounds for j in axes(B, 2), i in axes(A, 1)
+        @inbounds @simd for k in axes(A,2)
+            R[i,j] += A[i,k] * B[k,j]
+        end
+    end
+    nothing
+end
+
+# ╔═╡ 29ee6b7a-971e-4d46-ac3b-d793124a4542
+begin
+	A = rand(1024, 64)
+	B = rand(64, 512)
+
+	R = zeros(size(A,1), size(B,2))
+	∂z_∂R = rand(size(R)...)  # Some gradient/tangent passed to us
+
+	∂z_∂A = zero(A)
+	∂z_∂B = zero(B);
+end
+
+# ╔═╡ 113ca417-ede9-4f15-85ff-7c6fb7d97494
+Enzyme.autodiff(mymul!, 
+	Duplicated(R, ∂z_∂R),
+	Duplicated(A, ∂z_∂A),
+	Duplicated(B, ∂z_∂B))
+
+# ╔═╡ 62182140-d067-4814-a83d-f6c9d6b3a15f
+R ≈ A * B
+
+# ╔═╡ f6856069-8bfa-49ed-8f5e-f1bd6009c4f4
+∂z_∂A ≈ ∂z_∂R * B' # equivalent to Zygote.pullback(*, A, B)[2](∂z_∂R)[1]
+
```

Comment:
This explanation is a bit odd because it makes it seem like this manual math is equivalent to zygote, rathre than explaining the result of Enzyme sets the value the same as the pullback.